### ISSUE TITLE
OVH default imagePullPolicy is Always

### ISSUE THREAD

**Issue by minrk** [jupyterhub/mybinder.org-deploy#1786]
When we create resources without an imagePullPolicy, the default policy [should be IfNotPresent](https://kubernetes.io/docs/concepts/configuration/overview/#container-images) if the tag is

/cc @mael-le-gal how can we fix the default pull policy on OVH? Do you know where the source of the bug is? Can we fix it?

**Comment by minrk** [comment]
Turns out this is a deeper issue and #1787 does not fix it. `imagePullPolicy` does not appear to be respected at any level, as even setting `imagePullPolicy: IfNotPresent` in the hub deployment results in creating the hub pod with `imagePullPolicy: Always`. This means we can't bring OVH back until this is fixed at the cluster-level. This needs support from OVH folks.

**Comment by consideRatio** [comment]
Hmmm, I guessed there was a mutating webhook setup doing this, but there wasn't any except cert-managers stuff also in the other clusters (`kubectl get MutatingWebhookConfiguration`).

**Comment by consideRatio** [comment]
I googled on the docker pull rule and found...

> The rules are: Anonymous users can pull 100 images in six hours. Authenticated users can pull 200 images in six hours. Paid users are not limited.

Does it mean that if we have multiple IPs, it will be perceived as different anonymous users, and therefore we would get 100 images in six hours per machine, or, are we currently pulling docker images as the same authenticated user? Then I think we would get less capacity.

**Comment by minrk** [comment]
There is no mutating webhook.

We do have a docker credential for single-user images, at least. I don't know if it's being used for both init contianers and regular containers. I assume it's both. I'm not sure if each VM would have a different egress IP or not. All I know is that we were hitting the rate limit with very low traffic after turning OVH back on yesterday.

**Comment by consideRatio** [comment]
> I'm not sure if each VM would have a different egress IP or not.

On GKE, you have private and public k8s clusters, where public versions get an ip per node for egress traffic, while private will need to have a Cloud NAT setup. With a NAT to provide public IP(s) one can make it become one or more different IPs. But, this is GCP specific details.

**Comment by betatim** [comment]
I think we should wait and see what Mael has to say. It seems very likely that this behaviour is something that can be configured at the level of the cluster. So I wouldn't investigate workarounds until we know that it can't be changed.

**Comment by mael-le-gal** [comment]
Indeed I reproduced the problem easily.
As was previously said here : https://github.com/jupyterhub/mybinder.org-deploy/issues/1791 maybe we should first try to upgrade the k8s cluster because the current version is not supported anymore.
if the problem still persist, I would investigate with the kubernetes team about this strange behaviour.

**Comment by mael-le-gal** [comment]
I have a answer about the `imagePullPolicy` it is about they activate this feature by default : https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages

They are aware that it is a limitation for some users but for now we will have to deal with it...

The initial problem was about dockerhub rates right ? Back in the days we had a dedicated registry for binder. Would it fix the problem if I re-enable it for you ?

**Comment by betatim** [comment]
> The initial problem was about dockerhub rates right ? Back in the days we had a dedicated registry for binder. Would it fix the problem if I re-enable it for you ?

That would work. The reason we switched is because maintaining the registry was tricky. So we should see what we can do about that.

Can you explain who "they" are? I think you mean the admin team at OVH but not sure.

**Comment by minrk** [comment]
> Would it fix the problem if I re-enable it for you ?

Not entirely. It would help for builds, but we still have some images that come from docker hub, e.g. the network init container, binder, and hub pods images. The init container image is the main one causing the issue, so we could probably find a workaround for that one just on OVH if we really have to.

> they activate this feature by default

You said "by default," does that mean it can be changed? If not, we may need to keep OVH out of the federation indefinitely until it is addressed.

**Comment by mael-le-gal** [comment]
> The reason we switched is because maintaining the registry was tricky.

Yes I remember but at that time, our `registry as a service` was in beta. Stability is much better now.

> Can you explain who "they" are? I think you mean the admin team at OVH but not sure.

Sorry, yes I mean our team managing `kubernetes as a service`

> You said "by default," does that mean it can be changed?

No It is also forced on their side for now. I can ask them for a new "feature" but they will have to integrate my request on their roadmap and prioritize it, so I have no vision about when it will be available.

> Not entirely. It would help for builds, but we still have some images that come from docker hub, e.g. the network init container, binder, and hub pods images. The init container image is the main one causing the issue, so we could probably find a workaround for that one just on OVH if we really have to.

Yes I understand the problem. Can the image from dockerhub be overriden by configuration ? If yes we can push needed public image on our registry and use that registry as main.

**Comment by manics** [comment]
Could you setup your registry as a pull-through cache? https://docs.docker.com/registry/recipes/mirror/

**Comment by minrk** [comment]
> Can the image from dockerhub be overriden by configuration ?

#1805 should help with this. We'll need to customize the OVH deployment to run chartpress a second time with different configuration. That's more customization than I think a federation member should need, but it should be doable. I think we can make single-user images (including init containers) work without too much trouble. *Never* hitting docker hub is not as feasible, and I think not a fair requirement. Build images and chart dependency images in particular cannot as easily come from somewhere else. There are far fewer builds than launches, though, so this may not be so bad. If we run into the rate limit even after moving user images to a private registry, I think we would need to wait for the new feature before allowing OVH back into the federation.

>  I can ask them for a new "feature" but they will have to integrate my request on their roadmap and prioritize it, so I have no vision about when it will be available.

I think it is good to start this process, regardless of whether we are able to work around it for this particular deployment. This feature is based on the assumption that pulling an image you already have is cheap, which is not true anymore in a very large number of cases.

**Comment by minrk** [comment]
@mael-le-gal with #1805 merged, I think we can proceed with using an internal registry again. Can you deploy that and add the necessary credentials?

Tasks remaining to get OVH back up:

- [ ] deploy private registry
- [ ] add registry credentials, config to chart (imagePrefix)
- [ ] customize OVH CD to run chartpress against registry
- [ ] bring OVH back into federation

**Comment by mael-le-gal** [comment]
I just spawned a new custom registry and add creds on following PR : https://github.com/jupyterhub/mybinder.org-deploy/pull/1816

**Comment by minrk** [comment]
OVH is running images from chartpress in the private registry with #1821 and seems to be working. I'm at the end of my day, so I don't want to bring OVH back in just before the weekend, but I think we've done everything we need to to bring it back into rotation next week. Thanks @mael-le-gal!
